### PR TITLE
Simplifications and fixes to multicore systhreads implementation (3/3)

### PR DESCRIPTION
**Pull request by gadmm**
Follow-up to #11271, #11385 and #11393 (cc @Engil, @kayceesrk).

Targets 5.1 (does not seem to fix important bugs). Details in the individual commit logs. Can you please have a look at the commit that removes dead code, in case this was unintended and hides a bug? I also added a FIXME regarding a behaviour from old systhreads that is not yet implemented in multicore (see comments below).

This PR lives on top of #11393, so only the 5 topmost commits belong to it.

No change entry needed.

### CODE REVIEW

**Comment by gadmm** [comment]
I have rebased on top of #11393, so keep in mind that only the 5 topmost commits belong to this PR.

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [inline otherlibs/systhreads/st_stubs.c]: I restored this comment from old systhreads. However, this behaviour does not seem implemented yet in multicore.

**Comment by gadmm** [comment]
I rebased on top of #11385 and added some new minor fixes and clarifications. It still only contains less important fixes targeting 5.1. For clarity I convert it as a draft, we should review it when earlier PRs have been merged.

I added some more FIXMEs that I think are worth looking into.

(of course a round of precheck is welcome)

**Comment by gasche** [comment]
In case anyone wonders (I just did), the current "PR path" appears to be:

- already merged:
   + #11271, which was then reverted
   + #11390
  + #11403
- next to review:
   + #11385
- further PRs:
   + #11386 (the present PR)

**Comment by gasche** [comment]
(woopsie, the next one to review in #11403; edited)

**Comment by gadmm** [comment]
This should be rebased and made ready for review. While adapting to the new way systhread is initialized on every thread, I stumbled into a rabbit hole regarding error handling inside `Domain.spawn` and found other bugs with signals. I am not yet sure how to go about it, I just wanted to post an update.

**Comment by gadmm** [comment]
I have rebased on top of trunk. This could do with a round of review. Instructions for reviewers:
- This is best reviewed commit-per-commit, with the help of additional detail in each commit log.
- If someone knows systhread well, please review that commit "Remove dead code" does not reveal a regression in functionality compared to OCaml 4 (code that became dead for wrong reasons).
- A few FIXMEs are added, coming from my review of the OCaml 5 systhread implem. and are about possible regressions wrt. OCaml 4. I do not intend or expect to have to fix these FIXMEs in this PR.
  - Some FIXME concern possible changes/regressions wrt. the "best effort" to support `fork` on a single domain.
  - One FIXME is about the delay for joining the tick threads on shutdown (25ms on average). On OCaml 4 there was no wait. When reviewing the code, it appeared to me that a solution could be to interrupt the `sleep` system call in the tick thread using the self-pipe trick with `pselect` (edit: or simply using `pthread_cond_timedwait`).
- One FIXME is added to avoid delaying this PR further: the FIXME in `domain_thread_func` is about proper error handling inside domain initialization. Domain initialization was moved from OCaml to C code at the end of 5.0 development, which has created new issues. Errors during domain initialization should be converted into exceptions; currently this probably crashes or deadlocks. As said in my previous comment, fixing this has revealed to be a bit of a rabbit hole, it is doable but is better done separately in order to make progress on this PR. However, reviewers are free to disagree and refuse the addition of a FIXME comment in the code (in which case we can discuss solutions here).

**Comment by gasche** [comment]
Do we have a volunteer to review this PR? I have assigned myself as "person in charge of making sure we don't forget about it completely", but I would be happy to let someone else do the review.

**Comment by kayceesrk** [comment]
I'm starting to review this PR, but I have a couple of difficulties in the way this PR and others like #11307 are organised which is making it harder for me to review the PR(s). 

My main difficulty is that the PRs suggest many improvements that seem independent but are organised as several commits, each of which has to be read one after the other. The details of the improvements are in the commit messages. This has several downsides that make it harder to review such PRs.

* On Github, the discussions on several unrelated improvements get mixed together in the same PR thread which makes it harder to follow along.
* If some of the early improvements are OK, but the follow-ups need changes, the whole PR gets blocked. As a reviewer, I am not able to checkpoint my approval on individual commits. I'm forced to read the entire commit chain in order to see how these interact with each other.
* IINM, this is also different from how OCaml compiler development operates today, which is a PR for every improvement that's independent. 

Personally, I really find it hard to review PRs commit by commit. 

May I recommend @gadmm to consider having separate PRs for each improvement so that it makes reviewing easier? Small improvements will go much faster with a PR per commit. If they are inter-related, please mention that it builds on top of another PR. Also, possibly consider marking the PRs that are targets in a dependence chain as drafts so that reviewer time is not wasted on reviewing PRs which may change based on whether the source PRs require changes. Reviewer time is precious and limited, and it would help if the PRs are organised to be reviewer friendly.

File: otherlibs/systhreads/st_pthreads.h
Comment (kayceesrk) [inline otherlibs/systhreads/st_pthreads.h:124]: No change necessary, but noting that since we don't know which operation resulted in the non-`0` return value, all we can conclude is that `0` is success and non-`0` is an error. In particular, one cannot usefully inspect the error number.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c]: Is `0 != st_masterlock_init(m)` better than `st_masterlock_init(m) != 0`?

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c:409]: I can see that the proposed change is better than what was earlier. Is there no better recovery here than a `caml_fatal_error`?

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c]: This semantics is the opposite of the semantics of other thread operations like `pthread_mutex_init` and `pthread_cond_init` where `0` is success and non-zero is an error. Wondered whether there was a reason to do it this way?

File: runtime/domain.c
Comment (kayceesrk) [inline runtime/domain.c:1117]: Is the idea here that `domain_thread_func` is used as the `start_routine` for `pthread_create` and hence, it is not sensible for it to throw an exception? The `Domain.spawn` function catches all the synchronous exceptions. Are we  concerned with asynchronous exceptions here?

**Review by kayceesrk — COMMENTED** [review]
The first commit looks good to me. I've left some comments for optional improvements.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c:658]: The commit comment refers to `Thread_main_lock`. I can't find references to it. Has it been renamed to something else?

At a high-level, the invariant that we need is that `Active_thread` is non-NULL when `caml_c_thread_register` is called. The comment here says `Active_thread` may be NULL at safe point

https://github.com/ocaml/ocaml/blob/bbabb4b7075e49b5a5277633b5925e16c64f046e/otherlibs/systhreads/st_stubs.c#L138-L139

Is it not possible that `caml_c_thread_register` may be invoked at this point? 

(Meta comment: at this point, I have questions regarding the second commit, but I would be happy to have merged the first commit on its own.)

File: otherlibs/systhreads/thread.ml
Comment (kayceesrk) [inline otherlibs/systhreads/thread.ml:78]: This change looks good to me.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c:166]: This commit is fine.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c:525]: ~IIUC, the fix is to not join the systhread since it might be sleeping?~ I noticed that the commit messages confirms this.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c:619]: This commit LGTM.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [inline otherlibs/systhreads/st_stubs.c]: These are useful comments. This commit LGTM.

File: testsuite/tests/lib-unix/common/multicore_fork_domain_alone.ml
Comment (kayceesrk) [inline testsuite/tests/lib-unix/common/multicore_fork_domain_alone.ml:17]: This is much better. Thanks!

**Comment by kayceesrk** [comment]
Except for this one comment on the second commit (https://github.com/ocaml/ocaml/pull/11386/commits/142beb885999e07d5273fb8958811e5c077b0e9b#r1207685036), the 
other changes look fine to me.

**Review by kayceesrk — APPROVED** [review]

File: otherlibs/systhreads/st_pthreads.h
Comment (avsm) [reply otherlibs/systhreads/st_pthreads.h:124]: I think we can simplify this by thinking about which errors would be useful to ultimately signal to the mutator -- there can only really be `Out_of_memory` or another exception to signal that a thread cannot be created (to distinguish between memory exhaustion and other system limits like thread limits).

In this function, the only way the `pthread_mutex_init` or `pthread_mutex_cond` functions can fail is with a memory or resource error (`EAGAIN` or `ENOMEM`), since the functions aren't called in a way that could return `EINVAL`.  So just returning a negative error code of -1 to indicate an OOM condition would be simpler than propagating the results of one of the mutex calls failing, I think.

File: otherlibs/systhreads/st_stubs.c
Comment (avsm) [reply otherlibs/systhreads/st_stubs.c]: Could the comment be a little more specific about the actual precondition?  `caml_thread_reinitialize` is called as part of the `caml_atfork_hook`, and therefore this function is executed right after a `fork` (and then the rest of the proposed change makes more sense in context).

File: otherlibs/systhreads/st_stubs.c
Comment (avsm) [reply otherlibs/systhreads/st_stubs.c]: The house style in the runtime is mostly the latter, with only a few sprinklings of the `(0 == ...)` style that I can find in runtime/sys.c

File: otherlibs/systhreads/st_stubs.c
Comment (avsm) [reply otherlibs/systhreads/st_stubs.c:409]: Is even caml_fatal_error safe here, given the earlier giant "only execute async-signal-safe operations" comment in the diff hunk above?   If we've failed to do something as simple as allocate a few mutexes at this point in the fork, just exiting is probably the only option.

File: otherlibs/systhreads/st_stubs.c
Comment (avsm) [reply otherlibs/systhreads/st_stubs.c]: Agreed with @kayceesrk.   It's the opposite of the convention established in `st_masterlock_init` in this diff.

File: otherlibs/systhreads/st_stubs.c
Comment (avsm) [inline otherlibs/systhreads/st_stubs.c]: It would be really nice to distinguish between an OOM (`Out_of_memory`) and a thread allocation fail (`Failure`) here, if possible.

File: otherlibs/unix/fork.c
Comment (avsm) [inline otherlibs/unix/fork.c]:
```suggestion
      ("Unix.fork may not be called after any domain has been spawned");
```

is a more correct sounding perfect tense here.

File: otherlibs/unix/unix.mli
Comment (avsm) [inline otherlibs/unix/unix.mli]: This isn't entirely clear on how other threads have been spawned, given the invariant earlier. Do you mean system threads independently spawned of the OCaml runtime?

Comment (gasche) [reply otherlibs/unix/unix.mli]: Yes, I think this sentence refers to pthreads that would have been created by C code or other foreign code in the current process.

**Comment by gasche** [comment]
@gadmm it looks like the present PR has been reviewed three weeks ago, and would be mergeable in trunk after you rebase and take review comments into account. Are you available to do this in the next few weeks? If not, would you prefer if a maintainer took care of the rebase?

**Comment by gadmm** [comment]
As I understand it, this is not scheduled for 5.1 so there is no rush.

**Comment by kayceesrk** [comment]
> As I understand it, this is not scheduled for 5.1 so there is no rush.

It would be useful to get this towards a merge soon as the code has already gone out of sync with trunk and needs to be rebased. Merging it faster would mean that fixing conflicting changes would be the responsibility of folks who propose newer PRs ;-)

**Comment by gadmm** [comment]
Thanks for the gentle pressure, I am well aware and this has been at the top of my pile. As you can guess my own time is constrained, and I also received many solicitations from the OCaml github concurrently. I have been prioritizing 5.1. (Even though the release schedule is kept private for some odd reason, I have some sense that 5.1 is the priority at this time.)

Still on the topic of what we could do better to make sure that aging PRs end up getting merged, would monthly reminders be received positively? Would they help?

File: otherlibs/systhreads/st_pthreads.h
Comment (gadmm) [reply otherlibs/systhreads/st_pthreads.h:124]: It is true that you do not know which one failed, but e.g. in case of ENOMEM this does not really matter.

The docs also mention EBUSY and the function is indeed used to reinitialize a mutex after fork, so since what is done for fork+threads is not really standards-compliant, it is best to keep the explicit error message for debugging. After spending an hour trying to figure out the best course of action here, I am thinking we might be overthinking it. 

Raising an exception à la sync_check_error (which correctly maps ENOMEM to Out_of_memory) is doable, after fixing the handling of errors during domain creation.

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: Thanks, fixed. Comments about style are welcome, to help fix the inconsistency of the code base.

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:409]: Indeed if it fails, there is not much that can be done at that point. In case there is something to improve to the "best effort" to support fork+threads, it is still better to attempt writing an error message. Note that when trying to support fork+threads, one is already deep in unchartered territory (with some hope that the global lock ensures enough consistency to let you ignore the "only async-signal-safe" requirement). Of course, the whole enterprise is doomed to failure if there are other C threads and one is interrupted in the middle of malloc, say.

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: The reasoning was that the other ones are internal, while this one is exported in some sense (my view was probably influenced by hacking heavily the runtime hooks in the past).

The new version raises an exception instead. This is mostly useful when initializing threads on the first domain, which was already prepared to receive an exception.

When fixing domain_thread_func to handle errors, this will need to be changed to returning an encoded exception.

File: runtime/domain.c
Comment (gadmm) [reply runtime/domain.c:1117]: More work is to be done to correct handle exceptions here. In this case, an asynchronous exception can appear indeed. As discussed elsewhere, this is for another PR.

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: Since I feel that my comment was not very clear: I am backtracking a bit on returning an error code and I keep the exception-raising behaviour of caml_stat_alloc. I add a check to raise an exception on faulty mutex initialization. This lets me improve the behaviour when initializing threads on the first domain, without fixing or making worse the behaviour on initialization for other domains. This last fix will have to come in another PR (as mentioned already).

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:658]: Fixed `Thread_main_lock` -> `Thread_lock(Dom_c_threads)` in the commit description.

Roughly speaking, the `NULL` case for `Active_thread` is for before the threads are initialized on a domain, or during domain shutdown. But the code is not prepared to handle either case anyway, if it could ever happen. (Whether it should is a different question.)

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:525]: Note, as I mentioned in https://github.com/ocaml/ocaml/pull/11386#issuecomment-1456384769, that not joining is more problematic for multicore, and I suggested the use of pthread_cond_timedwait with a condition variable used to request continuing or shutting down as a possible solution.

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: Thanks, I took your phrasing verbatim.

File: otherlibs/systhreads/st_stubs.c
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: Wish granted. Note that properly dealing with those exceptions for domains ≥ 1 will be more work, as mentioned.

File: otherlibs/unix/fork.c
Comment (gadmm) [reply otherlibs/unix/fork.c]: Thanks

File: otherlibs/unix/unix.mli
Comment (gadmm) [reply otherlibs/unix/unix.mli]: Also systhreads, which is what I meant originally. C/FFI programmers already know not to mix fork & threads. I clarified the comment.

**Comment by gadmm** [comment]
The "fixup" comments are for reviewing my fixes. They will be squashed later.

**Comment by kayceesrk** [comment]
> Except for this one comment on the second commit (https://github.com/ocaml/ocaml/commit/142beb885999e07d5273fb8958811e5c077b0e9b#r1207685036), the
other changes look fine to me.

Force push lost my comment ?! :-( I don't know what my comment was, but I am assuming that it was fixed. If @gadmm can confirm, I think the PR is good to go.

**Comment by gadmm** [comment]
The comment is still there right above (https://github.com/ocaml/ocaml/pull/11386#discussion_r1207685036), along with my reply, but it looks like your pointer was invalidated. (I got the still-valid link by doing "copy link" directly from the PR discussion, which seems to give a different URL than yours.)

Let me know about the last point, and then I will clean-up the PR.

File: otherlibs/systhreads/st_stubs.c
Comment (kayceesrk) [reply otherlibs/systhreads/st_stubs.c:658]: Sounds good.

File: runtime/domain.c
Comment (kayceesrk) [reply runtime/domain.c:1117]: Ok.

**Comment by kayceesrk** [comment]
> Let me know about the last point, and then I will clean-up the PR.

LGTM. Please also add a changes entry. I'll aim to merge this right after that.

**Comment by gadmm** [comment]
Rebased & added a Changes entry (mostly to credit the work done by reviewers, as no change entry was needed for these minor changes & fixes IMO).

Edit: as already mentioned in a different discussion, the general advice on the organisation of this PR, which mentions various points that were already applied, and did not seem related to the delay in starting the review, ended up discussed in private.

**Pull request merged**
